src/upper_envelope/fues_jax/check_and_scan_funcs.py

```diff
@@ -0,0 +1,592 @@
+            point we scan from.
+        policy_to_scan_from (float): The policy function point to scan from. We want to
+            find the grid point which is on the same value function segment as the point
+            we scan from.
+        idx_to_scan_from (int): Index of the point we want to scan from. This should
+            be the current point we inspect.
+        endog_grid (np.ndarray): 1d array of shape (n_grid_wealth,) containing the
+            unrefined endogenous wealth grid.
+        value (np.ndarray): 1d array of shape (n_grid_wealth,) containing the
+            unrefined value correspondence.
+        policy (np.ndarray): 1d array of shape (n_grid_wealth,) containing the
+            unrefined policy correspondence.
+        n_points_to_scan (int): Number of points to scan.
+        jump_thresh (float): Threshold for the jump in the value function.
+        direction (str): The direction of the scan. Either 'forward' or 'backward'.
+
+    Returns:
+        tuple:
+
+        - grad_we_search_for (float): The gradient we search for.
+        - idx_on_same_value (int): The index of the point on the same value function
+            segment as the point we scan from.
+
+    """
+    # Prepare body function by partialing in, everything except carry and counter
+    partial_body = partial(
+        back_and_forward_scan_body,
+        endog_grid_to_calculate_gradient=endog_grid_to_calculate_gradient,
+        value_to_calculate_gradient=value_to_calculate_gradient,
+        endog_grid_to_scan_from=endog_grid_to_scan_from,
+        policy_to_scan_from=policy_to_scan_from,
+        endog_grid=endog_grid,
+        value=value,
+        policy=policy,
+        jump_thresh=jump_thresh,
+        direction=direction,
+    )
+
+    if direction == "forward":
+        max_index = idx_to_scan_from + n_points_to_scan
+
+        def cond_func(carry):
+            (
+                is_on_same_value,
+                idx_on_same_value,
+                grad_we_search_for,
+                current_index,
+            ) = carry
+            return (
+                is_scan_needed
+                & ~is_on_same_value
+                & (current_index < max_index)
+                & (current_index < len(endog_grid))
+            )
+
+        start_index = idx_to_scan_from + 1
+
+    elif direction == "backward":
+        min_index = idx_to_scan_from - n_points_to_scan
+
+        def cond_func(carry):
+            (
+                is_on_same_value,
+                idx_on_same_value,
+                grad_we_search_for,
+                current_index,
+            ) = carry
+            return (
+                is_scan_needed
+                & ~is_on_same_value
+                & (current_index > min_index)
+                & (current_index >= 0)
+            )
+
+        start_index = idx_to_scan_from - 1
+    else:
+        raise ValueError("Direction must be either 'forward' or 'backward'.")
```

Comment:
Do we need the ```direction``` check here? Isn't the the ```back_and_forward_scan_wrapper``` function supposed to be an internal function only?

Reply:
We define different condition functions, start indices and later update rules. So either we split the functions up in a more efficient way or use this flag if. I think it is only a minor compilation time difference but no definitely no computation time.